### ISSUE TITLE
Text classification by data compression  Max Halford

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9654]
Text classification by data compression • Max Halford<br>
<br>
https://ift.tt/v3dfmnJ<br>
<br>
Max Halford<br>
<br>

<div><div><div><div><div>
<p><strong>Edit</strong> – <em>I posted this <a href="https://news.ycombinator.com/item?id=27440093">on Hackernews</a> and got some valuable feedback. Many brought up the fact that you should be able to reuse the internal state of the compressor instead of recompressing the training data each time a prediction is made. There’s also some insightful references to data compression theory and its ties to statistical learning</em></p>
<p>Last night I felt like reading <a href="http://aima.cs.berkeley.edu/"><em>Artificial Intelligence: A Modern Approach</em></a>. I stumbled on something fun in the natural language processing chapter. The section I was reading dealt with classifying text. The idea of the particular subsection I was reading was to classify documents by using a <a href="https://www.wikiwand.com/en/Data_compression">compression algorithm</a>. This is such a left field idea, and yet it does make sense when you think about it. To quote the book:</p>
<blockquote><p>In effect, compression algorithms are creating a language model. The <a href="https://www.wikiwand.com/en/Lempel%E2%80%93Ziv%E2%80%93Welch">LZW algorithm</a> in particular directly models a maximum-entropy probability distribution.</p></blockquote>
<p>In other words, a compression algorithm has some knowledge of the distribution of words in a corpus. It can thus be used to classify documents. The learning algorithm is quite straightforward:</p>
<ol>
<li>Take a labeled training set of documents.</li>
<li>Build a single document per label by concatenating the texts that belong to that label.</li>
<li>Compress each obtained document and measure the size of each result.</li>
</ol>
<p>To classify a document, proceed as so:</p>
<ol>
<li>Concatenate the document with the concatenated document of each label.</li>
<li>Compress each of these concatenations and measure the size.</li>
<li>Return the label for which the size increased the least.</li>
</ol>
<p>The idea is that if a document is similar to the training texts associated with a particular label, they will share patterns that will get exploited by the compression algorithm. Ideally, the size increase of compressing the training texts with the new text should be correlated with the similarity between the text and the training texts. The smaller the increase, the more likely the label should be assigned.</p>
<p>I think this is an elegant idea. It’s not sophisticated, and I don’t expect it to perform better than a plain and simple logistic regression. Moreover, it’s expensive because the training texts of each label have to be recompressed for each test document. Still, I found the idea intriguing and decided to implement it in Python.</p>
<p>Here I’ll use the <a href="http://qwone.com/~jason/20Newsgroups/">newsgroup20 dataset</a> from scikit-learn. I’m using the same four categories they use in their <a href="https://scikit-learn.org/stable/datasets/real_world.html#converting-text-to-vectors">user guide</a> to have something to compare against.</p>
<div><pre tabindex="0"><code><span><span><span>from</span> <span>sklearn.datasets</span> <span>import</span> <span>fetch_20newsgroups</span>
</span></span><span><span>
</span></span><span><span><span>categories</span> <span>=</span> <span>[</span>
</span></span><span><span> <span>'alt.atheism'</span><span>,</span>
</span></span><span><span> <span>'talk.religion.misc'</span><span>,</span>
</span></span><span><span> <span>'comp.graphics'</span><span>,</span>
</span></span><span><span> <span>'sci.space'</span>
</span></span><span><span><span>]</span>
</span></span><span><span>
</span></span><span><span><span>train</span> <span>=</span> <span>fetch_20newsgroups</span><span>(</span>
</span></span><span><span> <span>subset</span><span>=</span><span>'train'</span><span>,</span>
</span></span><span><span> <span>categories</span><span>=</span><span>categories</span>
</span></span><span><span><span>)</span>
</span></span></code></pre></div>
<p>The first step is to concatenate the texts that belong to each of the four categories. I’m adding a space before each text so that the last word isn’t glued with the first word of the next text.</p>
<div><pre tabindex="0"><code><span><span><span>from</span> <span>collections</span> <span>import</span> <span>defaultdict</span>
</span></span><span><span>
</span></span><span><span><span>label_texts</span> <span>=</span> <span>defaultdict</span><span>(</span><span>str</span><span>)</span>
</span></span><span><span>
</span></span><span><span><span>for</span> <span>i</span><span>,</span> <span>text</span> <span>in</span> <span>enumerate</span><span>(</span><span>train</span><span>[</span><span>'data'</span><span>]):</span>
</span></span><span><span> <span>label</span> <span>=</span> <span>train</span><span>[</span><span>'target_names'</span><span>][</span><span>train</span><span>[</span><span>'target'</span><span>][</span><span>i</span><span>]]</span>
</span></span><span><span> <span>label_texts</span><span>[</span><span>label</span><span>]</span> <span>+=</span> <span>' '</span> <span>+</span> <span>text</span><span>.</span><span>lower</span><span>()</span>
</span></span></code></pre></div>
<p>The next step is to compress each of these big texts and measure the size of the compressed result. It’s quite easy to do this in Python. Indeed, Python provides high-level functions that compress a sequence of bytes into a smaller sequence of bytes. The <code>len</code> method gives us the number of bytes in the sequence. I picked <a href="https://docs.python.org/3/library/gzip.html"><code>gzip</code></a> at random from the compression methods listed <a href="https://docs.python.org/3/library/archiving.html">here</a>. Each of these provides an easy-to-use <a href="https://docs.python.org/3/library/gzip.html#gzip.compress"><code>compress</code></a> method.</p>
<div><pre tabindex="0"><code><span><span><span>import</span> <span>gzip</span>
</span></span><span><span>
</span></span><span><span><span>METHOD</span> <span>=</span> <span>gzip</span>
</span></span><span><span>
</span></span><span><span><span>original_sizes</span> <span>=</span> <span>{</span>
</span></span><span><span> <span>label</span><span>:</span> <span>len</span><span>(</span><span>METHOD</span><span>.</span><span>compress</span><span>(</span><span>text</span><span>.</span><span>encode</span><span>()))</span>
</span></span><span><span> <span>for</span> <span>label</span><span>,</span> <span>text</span> <span>in</span> <span>label_texts</span><span>.</span><span>items</span><span>()</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>print</span><span>(</span><span>original_sizes</span><span>)</span>
</span></span></code></pre></div>
<div><pre tabindex="0"><code><span><span><span>{</span>
</span></span><span><span> <span>'comp.graphics'</span><span>:</span> <span>252268</span><span>,</span>
</span></span><span><span> <span>'talk.religion.misc'</span><span>:</span> <span>224228</span><span>,</span>
</span></span><span><span> <span>'sci.space'</span><span>:</span> <span>310524</span><span>,</span>
</span></span><span><span> <span>'alt.atheism'</span><span>:</span> <span>266440</span>
</span></span><span><span><span>}</span>
</span></span></code></pre></div>
<p>That’s all there is to the training phase. The training texts have to be kept in memory because they have to be used for the prediction phase. Let’s say we’re given a list of unlabeled texts. The idea for each text is to concatenate it with each training text, compress the result, and then measure the size of the compressed result.</p>
<p>When that’s done, we just need to compare the obtained sizes with the original and return the label for which the size increased the least. Note that there is no notion of probability. There might be some weird way to cook up some probabilities, but they wouldn’t be <a href="https://scikit-learn.org/stable/modules/calibration.html">calibrated</a>.</p>
<div><pre tabindex="0"><code><span><span><span>test</span> <span>=</span> <span>fetch_20newsgroups</span><span>(</span><span>subset</span><span>=</span><span>'test'</span><span>,</span> <span>categories</span><span>=</span><span>categories</span><span>)</span>
</span></span><span><span><span>predictions</span> <span>=</span> <span>[]</span>
</span></span><span><span>
</span></span><span><span><span>for</span> <span>text</span> <span>in</span> <span>test</span><span>[</span><span>'data'</span><span>]:</span>
</span></span><span><span>
</span></span><span><span> <span>sizes</span> <span>=</span> <span>{</span>
</span></span><span><span> <span>label</span><span>:</span> <span>len</span><span>(</span><span>METHOD</span><span>.</span><span>compress</span><span>(</span><span>f</span><span>'</span><span>{</span><span>label_text</span><span>}</span><span> </span><span>{</span><span>text</span><span>.</span><span>lower</span><span>()</span><span>}</span><span>'</span><span>.</span><span>encode</span><span>()))</span>
</span></span><span><span> <span>for</span> <span>label</span><span>,</span> <span>label_text</span> <span>in</span> <span>label_texts</span><span>.</span><span>items</span><span>()</span>
</span></span><span><span> <span>}</span>
</span></span><span><span>
</span></span><span><span> <span>predicted_label</span> <span>=</span> <span>min</span><span>(</span>
</span></span><span><span> <span>sizes</span><span>,</span>
</span></span><span><span> <span>key</span><span>=</span><span>lambda</span> <span>label</span><span>:</span> <span>sizes</span><span>[</span><span>label</span><span>]</span> <span>-</span> <span>original_sizes</span><span>[</span><span>label</span><span>]</span>
</span></span><span><span> <span>)</span>
</span></span><span><span>
</span></span><span><span> <span>predictions</span><span>.</span><span>append</span><span>(</span><span>predicted_label</span><span>)</span>
</span></span></code></pre></div>
<p>So how well does this fair? Well, it takes over 5 minutes to run on my laptop for only 1,353 test cases. That’s 0.2 seconds per document, which is rather slow! Here’s the classification report:</p>
<div><pre tabindex="0"><code><span><span><span>from</span> <span>sklearn.metrics</span> <span>import</span> <span>classification_report</span>
</span></span><span><span>
</span></span><span><span><span>test_labels</span> <span>=</span> <span>[</span>
</span></span><span><span> <span>test</span><span>[</span><span>'target_names'</span><span>][</span><span>label</span><span>]</span>
</span></span><span><span> <span>for</span> <span>label</span> <span>in</span> <span>test</span><span>[</span><span>'target'</span><span>]</span>
</span></span><span><span><span>]</span>
</span></span><span><span>
</span></span><span><span><span>print</span><span>(</span><span>classification_report</span><span>(</span>
</span></span><span><span> <span>test_labels</span><span>,</span>
</span></span><span><span> <span>predictions</span><span>,</span>
</span></span><span><span> <span>digits</span><span>=</span><span>3</span>
</span></span><span><span><span>))</span>
</span></span></code></pre></div>
<pre tabindex="0"><code>                    precision    recall  f1-score   support

       alt.atheism      0.678     0.680     0.679       319
     comp.graphics      0.784     0.897     0.837       389
         sci.space      0.878     0.746     0.807       394
talk.religion.misc      0.609     0.614     0.611       251

          accuracy                          0.749      1353
         macro avg      0.737     0.734     0.733      1353
      weighted avg      0.754     0.749     0.749      1353
</code></pre>
<p>Is that good? No, not really. A multinomial Naive Bayes classifier achieves a macro F1-score of 0.88 with the same categories and train/test split. It’s also dramatically faster. Is there anything we can do? We can try other compression algorithms. Here are the results for <a href="https://docs.python.org/3/library/zlib.html"><code>zlib</code></a>, which was a bit faster and took just over 4 minutes to run:</p>
<pre tabindex="0"><code>                    precision    recall  f1-score   support

       alt.atheism      0.656     0.687     0.671       319
     comp.graphics      0.782     0.902     0.838       389
         sci.space      0.880     0.744     0.806       394
talk.religion.misc      0.612     0.578     0.594       251

          accuracy                          0.745      1353
         macro avg      0.732     0.728     0.727      1353
      weighted avg      0.749     0.745     0.744      1353
</code></pre>
<p>The performance with <code>zlib</code> seems to be worse than with <code>gzip</code>. What about <a href="https://docs.python.org/3/library/bz2.html"><code>bz2</code></a>?</p>
<pre tabindex="0"><code>                    precision    recall  f1-score   support

       alt.atheism      0.648     0.110     0.188       319
     comp.graphics      0.732     0.584     0.649       389
         sci.space      0.923     0.305     0.458       394
talk.religion.misc      0.271     0.928     0.420       251

          accuracy                          0.455      1353
         macro avg      0.644     0.482     0.429      1353
      weighted avg      0.682     0.455     0.442      1353
</code></pre>
<p>The performance is awful in comparison. Also, it took more than 7 minutes to run. Now how about <a href="https://docs.python.org/3/library/lzma.html"><code>lzma</code></a>?</p>
<pre tabindex="0"><code>                    precision    recall  f1-score   support

       alt.atheism      0.851     0.875     0.862       319
     comp.graphics      0.923     0.959     0.941       389
         sci.space      0.927     0.934     0.930       394
talk.religion.misc      0.866     0.773     0.817       251

          accuracy                          0.897      1353
         macro avg      0.892     0.885     0.888      1353
      weighted avg      0.897     0.897     0.896      1353
</code></pre>
<p>The performance is surprisingly good! But this comes at a cost: it took 32 minutes to complete, which is ~1.4 seconds per document. Still, it’s very interesting to see this kind of result for an algorithm that wasn’t at all designed to classify documents. Well, the <code>lzma</code> in fact implements the LZW algorithm that is mentioned in the <em>Artificial Intelligence: A Modern Approach</em> book, which might explain why it does so well!</p>
<p>There’s probably some fine-tuning that could be done to improve this kind of approach. However, it would most probably not be worth using such an approach because of the prohibitive computational cost. This should simply be seen as an interesting example of thinking outside the box. It also goes to show that statistical modelling and information theory are very much intertwined.</p>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live. <br>
<br>
July 14, 2023 at 11:00AM
